clang/lib/CodeGen/CGBlocks.cpp

```diff
@@ -1085,8 +1085,10 @@ llvm::Value *CodeGenFunction::EmitBlockLiteral(const CGBlockInfo &blockInfo) {
       blockAddr.getPointer(), ConvertType(blockInfo.getBlockExpr()->getType()));
 
   if (IsOpenCL) {
-    CGM.getOpenCLRuntime().recordBlockInfo(blockInfo.BlockExpression, InvokeFn,
-                                           result, blockInfo.StructureType);
+    CGM.getOpenCLRuntime().recordBlockInfo(
+        blockInfo.BlockExpression, InvokeFn, result, blockInfo.StructureType,
+        CurGD && CurGD.isDeclOpenCLKernel() &&
+            (CurGD.getKernelReferenceKind() == KernelReferenceKind::Kernel));
```

Comment:
Does that mean we replace the function body of the stub with a call to the respective kernel?
This way we emit the function body only once.
So a kernel to kernel function call is replaced by a kernel to stub to kernel call?

For example, the following snippet:
```
kernel void callee_kern(){

}
kernel void caller_kern(){
	callee_kern();
}
```

should generate IR that looks like:

```
define dso_local amdgpu_kernel void @callee_kern({{.*}}){
 
}
define dso_local void @_clang_ocl_kern_imp_callee_kern({{.*}}){
	entry:
	tail call amdgpu void @callee_kern({{.*}})
}
define dso_local amdgpu_kernel void @caller_kern({{.*}}){
	tail call void @__clang_ocl_kern_imp_callee_kern();
}
define dso_local void @_clang_ocl_kern_imp_caller_kern({{.*}}){
	tail call void @__clang_ocl_kern_imp_callee_kern();
}
```

Reply:
That's one of the options, yeah. You'd always emit stubs as directly calling their kernels (so `_clang_ocl_kern_imp_caller_kern` would actually just call `caller_kern`, at least coming out of IRGen), and then you'd let LLVM decide when and whether to inline. The main disadvantage is that there might not already be code to emit calls to kernels; I imagine that's normally only done by the OpenCL runtime, and the ABI is probably pretty different from the normal call ABI.  But maybe that's not a big deal.

The second option is that you could emit stubs by making sure the kernel is emitted first and then directly cloning the kernel body into the stub.  Whether this is materially different from just inlining, I don't know, but it's an option. We do have code for this sort of thing already — we need it for emitting virtual varargs thunks in C++.

The third option is that you could emit kernels as directly calling their stubs. The advantage here is that all the calls are just normal calls that you definitely already have code to handle. The disadvantage is that you'd always be forcing the stub to be emitted, even in the 99% case that there are no other uses of it. It'll probably get reliably inlined away, but still, it's burning some extra compile time.

The last option is that you could emit kernels by emitting the stub and then directly cloning the function body into the kernel.

Up to you; they all have trade-offs. But I do think you need to not double-emit the kernel body.

Reply:
> a kernel to kernel function call is replaced by a kernel to stub to kernel call 

Calls to kernel (from device and kernel functions) is not supported during backend IselLowering : https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp#L1127
So, we can't have a stub calling a kernel go into the backend.
In-lining the callee kernel body into the stub function might not be helpful in cases where callee kernel is just a function declaration.

> kernel is emitted first and then directly cloning the kernel body into the stub

> emit kernels by emitting the stub and then directly cloning the function body into the kernel

Some doubts on these points: Wouldn't this also result in having two copies of the kernel body ? Would this avoid the problems we see during double emission ?

Reply:
> Calls to kernel (from device and kernel functions) is not supported during backend IselLowering :

That makes sense.  Okay, so you need to call the stub.  That's probably fine — LLVM should reliably clean it up regardless — but it'll cause a lot of test case churn unless you force inlining prior to LLVM.

> Some doubts on these points: Wouldn't this also result in having two copies of the kernel body ? Would this avoid the problems we see during double emission ?

Having two copies of a function body is not the problem.  This is essentially just inlining, so as long as the kernel body doesn't do something that's semantically incompatible with inlining (which mostly just means the address-of-label extension), it's fine to inline.

The problem is that the visitor for emitting the body does not expect to run twice on the same code, and so it will emit code incorrectly for structures that require supporting global variables and functions, such as blocks and static locals.  For example, patch contains code to work around the double-emission of a block.